src/data-access/repositories/user-repository.ts

```diff
@@ -0,0 +1,197 @@
+import { UserInput } from '../../graphql/types/schema-types';
+import { UserModel, UserPageModel } from '../models/user';
+import AWS from 'aws-sdk';
+import { UserInputError } from 'apollo-server-errors';
+import moment from 'moment';
+import validator from 'validator';
+import { v4 as uuid } from 'uuid';
+import { ApolloError } from 'apollo-server-lambda';
+import { DocumentClient, QueryOutput, ScanOutput } from 'aws-sdk/clients/dynamodb';
+import { Maybe } from 'graphql/jsutils/Maybe';
+import { UsersDBConfiguration } from '../../configuration/users-db';
+
+export interface UserRepository {
+	createUser(data: UserInput): Promise<UserModel>;
+	updateUser(id: string, data: UserInput): Promise<UserModel>;
+	deleteUser(id: string): Promise<UserModel>;
+	getUser(id: string): Promise<UserModel>;
+	listUsers(query: Maybe<string>, limit: number, cursor: Maybe<string>): Promise<UserPageModel>;
+}
+
+interface ValidatedInput extends UserInput {
+	formattedDateOfBirth: string | null;
+}
+
+export class DynamoDBUserRepository implements UserRepository {
+	constructor(private database: AWS.DynamoDB.DocumentClient, private config: UsersDBConfiguration) {}
+
+	async createUser(data: UserInput): Promise<UserModel> {
+		if (!data.name) {
+			throw new UserInputError('The name must be specified');
+		}
+
+		if (!data.dob) {
+			throw new UserInputError('The date of birth must be specified');
+		}
+
+		if (!data.address) {
+			throw new UserInputError('The address must be specified');
+		}
+
```

Comment:
Is there another way to perform these validations?

Reply:
I could create a different Input for this resolver and require the parameters directly in the schema definition, but this would not address empty strings. I could define something like
`const requiredFields = ['name', 'dob', address]`
and then iterate on it to check for missing fields on the args object.

If you're talking about this app's validations as a whole, it would be better if we aggregated all errors in a list (or dictionary) before returning, that way we could return all the errors at once, and the client could fix them all in one go instead of fixing one at a time.